modules/reindex/src/internalClusterTest/java/org/elasticsearch/migration/FeatureMigrationIT.java

```diff
@@ -267,6 +270,62 @@ public void testMigrateIndexWithWriteBlock() throws Exception {
         });
     }
 
+    public void testMigrationWillRunAfterError() throws Exception {
+        createSystemIndexForDescriptor(INTERNAL_MANAGED);
+
+        TestPlugin.preMigrationHook.set((state) -> Collections.emptyMap());
+        TestPlugin.postMigrationHook.set((state, metadata) -> {});
+
+        ensureGreen();
+
+        SetOnce<Boolean> clusterStateUpdated = new SetOnce<>();
+        internalCluster().getCurrentMasterNodeInstance(ClusterService.class)
+            .submitStateUpdateTask(this.getTestName(), new ClusterStateUpdateTask() {
+                @Override
+                public ClusterState execute(ClusterState currentState) throws Exception {
+                    FeatureMigrationResults newResults = new FeatureMigrationResults(
+                        Collections.singletonMap(
+                            FEATURE_NAME,
+                            SingleFeatureMigrationResult.failure(INTERNAL_MANAGED_INDEX_NAME, new RuntimeException("it failed :("))
+                        )
+                    );
+                    Metadata newMetadata = Metadata.builder(currentState.metadata())
+                        .putCustom(FeatureMigrationResults.TYPE, newResults)
+                        .build();
+                    return ClusterState.builder(currentState).metadata(newMetadata).build();
+                }
+
+                @Override
+                public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+                    clusterStateUpdated.set(true);
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    clusterStateUpdated.set(false);
+                }
+            }, ClusterStateTaskExecutor.unbatched());
+
+        assertBusy(() -> assertTrue(Optional.ofNullable(clusterStateUpdated.get()).orElse(false)));
+
+        PostFeatureUpgradeRequest migrationRequest = new PostFeatureUpgradeRequest();
+        PostFeatureUpgradeResponse migrationResponse = client().execute(PostFeatureUpgradeAction.INSTANCE, migrationRequest).get();
+        // Make sure we actually started the migration
+        final Set<String> migratingFeatures = migrationResponse.getFeatures()
```

Comment:
Converting to a set isn't anymore efficient than just looking through the list. Is there som other reason we need an intermediate set to assert the feature was processed?

Reply:
This was a copy/paste from another test, you're right there's not much point to converting to a set here. I'll simplify.